### PR TITLE
Fix UTF32toUTF8 will produce invalid transition

### PR DESCRIPTION
**Pull request by tang-hi**
FIX ISSUE #12458

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java [hunk 1]
```diff
@@ -238,6 +238,10 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
         // doesn't accept certain byte sequences) -- there
         // are other cases we could optimize too:
         startCode = 194;
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:240]: Let's maybe fix this one to hex as well (0xC2)?

File: lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java [hunk 2]
```diff
@@ -238,6 +238,10 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
         // doesn't accept certain byte sequences) -- there
         // are other cases we could optimize too:
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:239]: Is this comment (`there are other cases we could optimize too`) still true :)  Or are these two new ifs covering them AND fixing this sneaky bug?

File: lucene/core/src/test/org/apache/lucene/util/TestUnicodeUtil.java [hunk 3]
```diff
@@ -188,6 +191,30 @@ public void testUTF8CodePointAt() {
     }
   }
 
+  public void testUTF8TwoToThreeBytes() throws Exception {
```
Comment (mikemccand) [inline lucene/core/src/test/org/apache/lucene/util/TestUnicodeUtil.java:194]: Maybe also add the three-to-four and one-to-two cases?

**Review by mikemccand — COMMENTED** [review]
Thanks @tang-hi -- what a sneaky bug!  It's crazy it took randomized testing so long to eek out this failure, and also wonderful that it did.

How do we characterize the end user impact of this bug?  We need a well written CHANGES entry.  The impact should be small, since it's accepting illegal UTF-8 sequences?  It'd only affect users operating entirely in binary space (not Unicode encoded UTF-8 terms, which is the vast majority of users / default case).

**Comment by tang-hi** [comment]
@mikemccand Thanks for the review. I have already made the updates to the pull request based on your review comments. 😄

**Comment by mikemccand** [comment]
> @mikemccand Thanks for the review. I have already made the updates to the pull request based on your review comments. 😄

Wow that was fast, thank you!  I'll try to review soon -- I gotta stare at this long enough to understand the bug / fix / UTF-8 details.  Thank you for digging on this tricky case so quickly.

**Comment by tang-hi** [comment]
2 byte -> 3 byte
image

3 byte -> 4 byte
image

hope these pictures will be helpful

File: lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java [hunk 4]
```diff
@@ -232,12 +232,18 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
           endUTF8.byteAt(upto) & (~MASKS[endUTF8.numBits(upto) - 1]),
           endUTF8.byteAt(upto));
     } else {
+      // There are three special case
+      // 1. if codepoint's numBits is 5, byte start from 0xC2
+      // 2. if codepoint's len is 3  and  first byte is 0xE0. the second byte start from 0xA0
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:237]: Extra space before `and` and `first`?

File: lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java [hunk 5]
```diff
@@ -232,12 +232,18 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
...
+      // you could found the reference in https://www.utf8-chartable.de/unicode-utf8-table.pl
       final int startCode;
       if (endUTF8.numBits(upto) == 5) {
-        // special case -- avoid created unused edges (endUTF8
-        // doesn't accept certain byte sequences) -- there
-        // are other cases we could optimize too:
-        startCode = 194;
+        startCode = 0xC2;
+      } else if (endUTF8.len == 3 && upto == 1 && endUTF8.byteAt(0) == 0xE0) {
+        startCode = 0xA0;
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:244]: Can you move the above three bullets into each of these internal `if`s to make the special casing clearer?

E.g. in this `if` maybe say:

```
// the first length=3 UTF8 Unicode character is E0 A0 80 so we must special case A0 as the 2nd byte when E0 was the first byte of endUTF8 in this case
```

**Review by mikemccand — COMMENTED** [review]
Phew, I finally understand this bug!

And I realize how I messed this up long ago after studying the UTF-8 encoding table ... from the line for the first 3-byte UTF-8 sequence:

```
U+0800 | U+FFFF | 1110xxxx | 10xxxxxx | 10xxxxxx
```

one cannot assume that the `xxxx` of the not-first UTF-8 byte will be 0s!  For the 2->3 and 3->4 case they are not :)

Thank you @tang-hi for bearing with me as I remember this tricky stuff.  I left a bunch of small comments to help readability maybe.  I think the fix is fundamentally correct!

File: lucene/core/src/test/org/apache/lucene/util/automaton/TestStringsToAutomaton.java [hunk 6]
```diff
@@ -142,22 +141,11 @@ private void checkAutomaton(List<BytesRef> expected, Automaton a, boolean isBina
     }
 
     // Make sure every term produced by the automaton is expected
-    FiniteStringsIterator it = new FiniteStringsIterator(a);
```
Comment (mikemccand) [inline lucene/core/src/test/org/apache/lucene/util/automaton/TestStringsToAutomaton.java:145]: Hmm why did you remove the `isBinary` case for this test?

File: lucene/core/src/test/org/apache/lucene/util/TestUnicodeUtil.java [hunk 7]
```diff
@@ -188,6 +191,37 @@ public void testUTF8CodePointAt() {
...
+    // utf8 codepoint length is 1
+    b.addTransition(s1, s2, 0x7F);
+    // utf8 codepoint length is 2
+    b.addTransition(s1, s2, 0x80);
+    b.addTransition(s1, s2, 0x7FF);
+    // utf8 codepoint length is 3
+    b.addTransition(s1, s2, 0x800);
+    b.addTransition(s1, s2, 0xFFFF);
+    // utf8 codepoint length is 4
+    b.addTransition(s1, s2, 0x10000);
```
Comment (mikemccand) [inline lucene/core/src/test/org/apache/lucene/util/TestUnicodeUtil.java:212]: I'm pretty sure the Automaton builder collapses adjacent transitions like this, but for paranoia, could you add the range explicitly?  E.g.:

```
b.addTransition(s1, s2, 0x7f, 0x80);
```

(Instead of two separate `addTransition` calls).  And same for the other two (four) transitions?

File: lucene/core/src/test/org/apache/lucene/util/automaton/TestStringsToAutomaton.java (hunk 6 above)
Comment (tang-hi) [reply lucene/core/src/test/org/apache/lucene/util/automaton/TestStringsToAutomaton.java:145]: Because it was added by @gsmiller in https://github.com/apache/lucene/pull/12461 for the purpose of passing the CI test, it should be restored since we have already fixed the bug.

**Comment by tang-hi** [comment]
I have already updated the PR. You may review it at your convenience.

**Comment by gsmiller** [comment]
Thanks @tang-hi for this change! I'll also try to have a look at this tomorrow or Friday since I already spent some time untangling the code in here :)

File: lucene/CHANGES.txt [hunk 8]
```diff
@@ -182,6 +182,10 @@ Bug Fixes
 * GITHUB#12451: Change TestStringsToAutomaton validation to avoid automaton conversion bug discovered in GH#12458
   (Greg Miller).
 
+* GITHUB#2472: UTF32ToUTF8 would sometimes accept extra invalid UTF-8 binary sequences.  This should not have any
```
Comment (gsmiller) [inline lucene/CHANGES.txt:185]: Unless a user was directly using `UTF32ToUTF8#convert` right? Maybe you could mention that somehow?

File: lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java [hunk 9]
```diff
@@ -35,12 +35,12 @@ public final class UTF32ToUTF8 {
...
   private static final int[] endCodes = new int[] {127, 2047, 65535, 1114111};
 
-  static int[] MASKS = new int[32];
+  static int[] MASKS = new int[8];
 
   static {
     int v = 2;
-    for (int i = 0; i < 32; i++) {
-      MASKS[i] = v - 1;
+    for (int i = 0; i < 7; i++) {
```
Comment (gsmiller) [inline lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:42]: Nice tidy-up!

File: lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java [hunk 10]
```diff
@@ -227,19 +227,24 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
...
-          endUTF8.byteAt(upto));
+          start, end, endUTF8.byteAt(upto) & (~MASKS[endUTF8.numBits(upto)]), endUTF8.byteAt(upto));
     } else {
       final int startCode;
-      if (endUTF8.numBits(upto) == 5) {
-        // special case -- avoid created unused edges (endUTF8
-        // doesn't accept certain byte sequences) -- there
-        // are other cases we could optimize too:
-        startCode = 194;
+      if (endUTF8.len == 2) {
```
Comment (gsmiller) [inline lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:233]: Maybe we should reference the issue number in a comment here that tracked the bug for future readers? It's a pretty tricky bug, so having the issue number in a comment might be useful?

File: lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java [hunk 11]
```diff
@@ -227,19 +227,24 @@ private void end(int start, int end, UTF8Sequence endUTF8, int upto, boolean doA
...
+        assert upto == 0; // the upto==1 case will be handled by the first if above
+        // the first length=2 UTF8 Unicode character is C2 80,
+        // so we must special case 0xC2 as the 1st byte.
+        startCode = 0xC2;
+      } else if (endUTF8.len == 3 && upto == 1 && endUTF8.byteAt(0) == 0xE0) {
+        // the first length=3 UTF8 Unicode character is E0 A0 80,
+        // so we must special case 0xA0 as the 2nd byte when E0 was the first byte of endUTF8.
+        startCode = 0xA0;
+      } else if (endUTF8.len == 4 && upto == 1 && endUTF8.byteAt(0) == 0xF0) {
```
Comment (gsmiller) [inline lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:242]: I'm sort of contorting my brain to try to figure out if this can ever actually happen since the max UTF8 length is 4 bytes. I'm not opposed to keeping it, but for my own sanity, did you actually trip this problem with a test case or see something I'm overlooking?

**Review by gsmiller — COMMENTED** [review]
Pretty elegant fix for a tricky bug. Thank you!

File: lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java (hunk 11 above)
Comment (tang-hi) [reply lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:236]: If you comment it out, the test `testUTF8SpanMultipleBytes` in `TestUnicodeUtil.java` will fail. This is because when there is a transition span from 0xFFFF (3 bytes) to 0x10000 (4 bytes), it will produce an incorrect result.

**Comment by tang-hi** [comment]
Hi, @gsmiller! I have already updated the PR. You may review it at your convenience.

**Comment by gsmiller** [comment]
Thanks @tang-hi. This LGTM now (I have two outstanding, small bits of feedback/questions above if you have a chance, but they're not blocking). I'll leave this open for another day incase Mike has any additional feedback, but will otherwise get it merged in a day or so. Thanks again!

File: lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java (hunk 11 above)
Comment (gsmiller) [reply lucene/core/src/java/org/apache/lucene/util/automaton/UTF32ToUTF8.java:236]: Right, of course. That makes sense. My brain was a bit tired at the end of the day yesterday when I was looking through this, and I had an "off by one" bug.

**Review by gsmiller — APPROVED** [review]
LGTM! Will merge tomorrow unless Mike has additional feedback.

**Comment by gsmiller** [comment]
Merged and back-ported. Thanks again @tang-hi! Nice fix for a reasonably tricky issue!
